src/backend/fs.rs

```diff
@@ -3,151 +3,127 @@
     let sk = SYS_CFG.private_key;
-    let ss = SharedSecret::new(&pk.into_inner(), &sk);
-    let write_pk = PublicKey::from_secret_key_global(&SecretKey::from_slice(&ss.secret_bytes())?);
+    let ss = SharedSecret::new(&pk.into_inner(), &sk).secret_bytes();
+    let write_pk = PublicKey::from_secret_key_global(&SecretKey::from_slice(&ss)?);
 
     let pk_bytes = write_pk.serialize();
     let (x_only_pk, _) = write_pk.x_only_public_key();
 
-    let file_hash = Blake3Hash(blake3::keyed_hash(&x_only_pk.serialize(), file_bytes));
-
-    trace!("TODO: Check if file catalog already exists");
-    // let path = SYS_CFG
-    //     .volumes
-    //     .get(0)
-    //     .expect("First volume present")
-    //     .path
-    //     .join(CATALOG_DIR)
-    //     .join(file_hash.to_string());
+    trace!("Initialize Blake3 keyed hasher");
+    let mut file_hasher = blake3::Hasher::new_keyed(&x_only_pk.serialize());
 
-    // trace!("Read catalog at {}", &path.to_string_lossy());
+    trace!("Iterate through file body stream");
+    let segment_hashes = file_stream
+        .map(|segment: Result<Bytes>| {
+            let segment = segment?;
+            file_hasher.update(&segment);
+            let encoded_segment = carbonado::encode(&pk_bytes, &segment, NODE_FORMAT)?;
+            let segment_hash = write_segment(&ss, &pk_bytes, &encoded_segment)?;
 
-    // let is_catalog = std::path::Path::new(&path).file_name();
-    // trace!("Catalog File: file_name: {:?}", is_catalog);
-    // if is_catalog.is_some() {
-    //     return Err(anyhow!("This file already exists for this user."));
-    // }
-
-    trace!("Segment files");
-    let segments_iter = file_bytes.par_chunks_exact(SEGMENT_SIZE);
+            Ok::<BaoHash, Error>(segment_hash)
+        })
+        .try_collect::<Vec<BaoHash>>()
+        .await?;
 
-    trace!("Encode each segment");
-    let remainder_bytes = segments_iter.remainder();
-    let last_segment = carbonado::encode(&pk_bytes, remainder_bytes, NODE_FORMAT)?;
+    let file_hash: Blake3Hash = Blake3Hash(file_hasher.finalize());
 
-    let mut encoded_segments = segments_iter
-        .map(|segment| carbonado::encode(&pk_bytes, segment, NODE_FORMAT))
-        .collect::<Result<Vec<Encoded>>>()?;
+    trace!("Check if segment already exists");
+    let path = SYS_CFG
+        .volumes
+        .get(0)
+        .expect("First volume present")
+        .path
+        .join(CATALOG_DIR)
+        .join(file_hash.to_string());
 
-    encoded_segments.push(last_segment);
+    trace!("Read catalog at {path:?}");
 
-    trace!("Get eight storage volumes from config");
-    if SYS_CFG.volumes.len() != FEC_M {
-        return Err(anyhow!("Eight volume paths must be configured"));
+    if path.exists() {
+        return Err(anyhow!("This file already exists for this public key."));
     }
 
-    trace!("Split each segment out into 8 separate chunks and write each chunk to the storage volume by filename");
```

Comment:
In the case of file_stream, how many separate chunks could there be, since we don't have a way to divide chunks.
I am supposing that the number of chunks is indeterminate.

Reply:
Correct, and it's a stream, so it cannot be known ahead of time.